users/models.py

```diff
@@ -1,3 +1,290 @@
+        return self.roles.exists() and self.roles.count() > 1
+
+    def soft_delete(self):
+        """
+        Soft delete the user account.
+
+        This method deactivates the account, sets the deleted_at timestamp,
+        and marks the account as deleted. Original user data is preserved.
+
+        Raises:
+            ValidationError: If the account is already inactive or deleted.
+        """
+        if not self.is_active or self.is_soft_deleted:
+            raise ValidationError("This account is already inactive or deleted.")
+
+        original_data = {
+            'email': self.email,
+            'first_name': self.first_name,
+            'last_name': self.last_name,
+            'user_phone': self.user_phone,
+            'about_me': self.about_me,
+        }
+        self.original_data = json.dumps(original_data)
+
+        self.email = f"deleted_{self.id}@example.com"
+        self.first_name = "Deleted"
+        self.last_name = "User"
+        self.user_phone = ""
+        self.about_me = ""
+        self.profile_picture.delete(save=False)
+        self.is_active = False
+        self.is_soft_deleted = True
+        self.deleted_at = timezone.now()
+        self.save()
+        logger.info(f"User {self.id} soft deleted")
+
+    def reactivate(self):
+        """
+        Reactivate a soft-deleted user account.
+
+        This method reactivates the account and restores the original user data.
+
+        Raises:
+            ValidationError: If the account is already active or not soft deleted.
+        """
+        if self.is_active or not self.is_soft_deleted:
+            raise ValidationError("This account is already active or not soft deleted.")
+
+        if self.original_data:
+            original_data = json.loads(self.original_data)
+            self.email = original_data['email']
+            self.first_name = original_data['first_name']
+            self.last_name = original_data['last_name']
+            self.user_phone = original_data['user_phone']
+            self.about_me = original_data['about_me']
+            self.original_data = None
+
+        self.is_active = True
+        self.is_soft_deleted = False
+        self.deleted_at = None
+        self.save()
+        logger.info(f"User {self.id} reactivated")
+
+    def save(self, *args, **kwargs):
+        """
+        Save the user instance to the database.
+
+        This method overrides the default save method to ensure
+        the email is always normalized before saving.
+
+        Args:
+            *args: Variable length argument list.
+            **kwargs: Arbitrary keyword arguments.
+        """
+        self.email = self.email.lower().strip()
+
+        if User.objects.filter(email=self.email).exclude(pk=self.pk).exists():
+            raise ValidationError("Email already exists.")
+
+        super().save(*args, **kwargs)
+
+    class Meta:
+        unique_together = [('email', 'is_active')]
```

Comment:
You are using `unique_together`. However, `unique_together` has been deprecated in favor of `UniqueConstraint` in Django 3.x. Refactor this constraint to use `UniqueConstraint`.
_Example_
```
constraints = [
        models.UniqueConstraint(fields=['email', 'is_active'], name='unique_active_email')
    ]
```